src/main/starlark/core/compile/cli/compile.bzl

```diff
@@ -0,0 +1,126 @@
+    if not srcs:
+        # still gotta create the jars for keep bazel haps.
+        actions.symlink(
+            output = output_srcjar,
+            target_file = toolchain_info.empty_jar,
+        )
+        actions.symlink(
+            output = class_jar,
+            target_file = toolchain_info.empty_jar,
+        )
+        return
+
+    classpath = depset(
+        transitive = [
+            dep_jars,
+            toolchain_info.kotlin_stdlib.transitive_compile_time_jars,
+        ],
+    )
+
+    # Set the stdlib for the compiler to run on.
+    args = actions.args().add("--wrapper_script_flag=--main_advice_classpath=%s" % (
+        ":".join([f.path for f in toolchain_info.kotlin_stdlib.transitive_compile_time_jars.to_list()])
+    ))
+    args.add("-d", class_jar)
+    args.add("-jdk-home", toolchain_info.java_runtime.java_home)
+    args.add("-jvm-target", toolchain_info.jvm_target)
+    args.add("-no-stdlib")
+    args.add("-verbose")
+    args.add("-api-version", toolchain_info.api_version)
+    args.add("-language-version", toolchain_info.language_version)
+    args.add("-module-name", module_name)
+    args.add_joined("-cp", classpath, join_with = path_separator)
+    for (k, v) in kotlinc_opts.items():
+        args.add(k, v)
+    args.add_all(srcs)
+
+    actions.run(
+        outputs = [class_jar],
+        executable = toolchain_info.kotlinc,
+        inputs = depset(direct = srcs, transitive = [classpath, toolchain_info.java_runtime.files]),
+        arguments = [args],
+        mnemonic = toolchain_info.compile_mnemonic,
+        toolchain = TYPE,
+    )
+
+    if output_srcjar:
+        actions.run(
+            outputs = [output_srcjar],
+            executable = toolchain_info.executable_zip,
+            inputs = srcs,
+            arguments = [actions.args().add("c").add(output_srcjar).add_all(srcs)],
+            mnemonic = "SourceJar",
+            toolchain = TYPE,
+        )
+
+def write_jvm_launcher(toolchain_info, actions, path_separator, workspace_prefix, jvm_flags, runtime_jars, main_class, executable_output):
+    template = toolchain_info.java_stub_template
+    java_runtime = toolchain_info.java_runtime
+    java_bin_path = java_runtime.java_executable_runfiles_path
+
+    # Following https://github.com/bazelbuild/bazel/blob/6d5b084025a26f2f6d5041f7a9e8d302c590bc80/src/main/starlark/builtins_bzl/bazel/java/bazel_java_binary.bzl#L66-L67
+    # Enable the security manager past deprecation.
+    # On bazel 6, this check isn't possible...
+    if getattr(java_runtime, "version", 0) >= 17:
+        jvm_flags = jvm_flags + " -Djava.security.manager=allow"
+
+    classpath = path_separator.join(
+        ["${RUNPATH}%s" % (j.short_path) for j in runtime_jars.to_list() + toolchain_info.kotlin_stdlib.transitive_compile_time_jars.to_list()],
+    )
+    needs_runfiles = "0" if java_bin_path.startswith("/") or (len(java_bin_path) > 2 and java_bin_path[1] == ":") else "1"
+
+    actions.expand_template(
```

Comment:
Added for completeness of the toolchain. If you want complete freedom to play with the kotlin compiler, you could use this.

Basically, it's a thin wrapper around kotlinc -- it can't support persistent workers (at least I don't see an easy path on it) -- but it will always be forward compatible (until they change the cli interface... I give it two major versions). It's the "no warranty" rules and toolchain for the discerning kotlin developer. I've even added basic plugin support: https://github.com/bazelbuild/rules_kotlin/pull/1251